notebooks/k_means.livemd

```diff
@@ -0,0 +1,559 @@
+  |> Vl.encode_field(:y, "petal_width", type: :quantitative)
+)
+```
+
+Now we have a better understanding of data. Iris species have different petal widths and petal lengths. Iris *Setosa* has the smallest petal *Versicolor* is medium size and *Virginica* has the biggest petal. We can make sure even further that our analysis is correct and plot the so-called **Elbow plot**. The Elbow plot is a plot which presents Inertia vs the number of clusters. If there is a characteristic elbow on, then we have a strong suggestion about the right number of clusters. Let's train KMeans models for a different number of clusters from range 1 to 11.
+
+```elixir
+models = for i <- 1..11, do: Scholar.Cluster.KMeans.fit(df, num_clusters: i)
+inertia_of_models = for model <- models, do: Nx.to_number(model.inertia)
+```
+
+```elixir
+data = [num_clusters: 1..11, inertia: inertia_of_models]
+
+Vl.new(width: 600, height: 300, title: "Elbow Plot")
+|> Vl.data_from_values(data)
+|> Vl.mark(:line)
+|> Vl.encode_field(:x, "num_clusters",
+  title: "Number of Clusters",
+  type: :ordinal,
+  axis: [label_angle: 0]
+)
+|> Vl.encode_field(:y, "inertia", title: "Inertia", type: :quantitative)
+```
+
+As you can see, we have the elbow when the number of clusters equals three. So this value of the parameter seems to be the best.
+
+```elixir
+best_model = Enum.at(models, 2)
+best_model = Datasets.sort_clusters(best_model)
+accuracy = Scholar.Metrics.accuracy(best_model.labels, y)
+```
+
+Accuracy is nearly 90%, that pretty decent! Let's look at our results plotted on one of the previous plots.
+
+```elixir
+coords = best_model.clusters[[0..-1//1, 2..-1//1]] |> Nx.to_flat_list()
+
+coords = [
+  x_coords: for(i <- [0, 2, 4], do: Enum.at(coords, i)),
+  y_coords: for(i <- [1, 3, 5], do: Enum.at(coords, i))
+]
+```
+
+```elixir
+Vl.new(
+  width: 300,
+  height: 300,
+  title: [
+    text:
+      "Scatterplot of data samples pojected on plane petal_width x petal_length with calculated centroids",
+    offset: 25
+  ]
+)
+|> Vl.layers([
+  Vl.new()
+  |> Vl.data_from_values(df_raw)
+  |> Vl.mark(:circle)
+  |> Vl.encode_field(:x, "petal_length", type: :quantitative)
+  |> Vl.encode_field(:y, "petal_width", type: :quantitative)
+  |> Vl.encode_field(:color, "species"),
+  Vl.new()
+  |> Vl.data_from_values(coords)
+  |> Vl.mark(:circle, color: :green, size: 100)
+  |> Vl.encode_field(:x, "x_coords", type: :quantitative)
+  |> Vl.encode_field(:y, "y_coords", type: :quantitative)
+])
+```
+
+As we expect 😎
+
+## Clustering of pixels' colors
+
+The other interesting use case of KMeans clustering is pixel clustering. This technique replaces all pixels with similar colors (similar in terms of euclidean distance between RGB) with a centroid related to them.
+
+<!-- livebook:{"break_markdown":true} -->
+
+Let us start with loading the referral image.
+
+```elixir
+{:ok, img} = StbImage.read_file("../scholar/test/data/taj_mahal.jpg")
```

Comment:
```suggestion
{:ok, img} = StbImage.read_file(Path.join(__DIR__, "../test/data/taj_mahal.jpg"))
```

Reply:
@msluszniak are there reliable locations we can download the image and the csv file from? Ideally we would download them with `req`, so the notebook runs without depending on local files.

Reply:
The CSV data is available [here](https://cam-orl.co.uk/facedatabase.html), with the image I think I should change it, cause I cannot find the usage rights.

Reply:
@jonatanklosko maybe it will be better to add the Olivetti dataset to [Scidata](https://github.com/elixir-nx/scidata) module. What do you think?

Reply:
Sounds good to me!

Reply:
@msluszniak yes, please do send a PR to Scidata. It also has some image datasets if you would prefer to use that instead. :)

Reply:
@josevalim unfortunately there is a problem with Olivetti. The data source (sci-kit uses it; [link](https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/datasets/_olivetti_faces.py#L32)) is saved as a .mat file. It would be problematic to parse this file into CSV. I think that I use fashionmnist instead.

Reply:
Ah, if fashionmnist can work that's great!